api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto

```diff
@@ -39,4 +40,60 @@ message ZooKeeperProxy {
   //
   // if that is set. If it isn't, ZooKeeper's default is also 1Mb.
   google.protobuf.UInt32Value max_packet_bytes = 3;
+
+  // List of latency thresholds for opcodes to decide the fast/slow responses and emit metrics (used for error budget calculation).
+  //
+  // https://sre.google/workbook/implementing-slos/
+  //
+  // If no latency threshold is set, the fast/slow response counters will always be 0.
+  // If the default threshold is not set, it will be set as 100 millisecond.
+  // If the threshold of one opcode is not set, it will fallback to use the default
+  // threshold.
+  //
+  // Specifying latency thresholds multiple times for one opcode is not allowed.
+  repeated LatencyThreshold latency_thresholds = 4 [(validate.rules).repeated = {max_items: 27}];
+}
+
+message LatencyThreshold {
+  enum Opcode {
+    Default = 0;
+    Connect = 1;
```

Comment:
Not a ZooKeeper expert, but this doesn't appear to be defined in [ZooDefs.java](https://github.com/apache/zookeeper/blob/89c1831f84891f425f1fa9224210587124f1c1ec/zookeeper-server/src/main/java/org/apache/zookeeper/ZooDefs.java#L36), and some of the other opcodes' values are shifted by +1.
Is this deviation necessary, or can we keep the them in sync?

Reply:
Unfortunately, we cannot keep them in sync since there are [some negative opcode values](https://github.com/apache/zookeeper/blob/89c1831f84891f425f1fa9224210587124f1c1ec/zookeeper-server/src/main/java/org/apache/zookeeper/ZooDefs.java#L98-L102) that cannot happen in proto.
Specifically, I followed the opcode order [here](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/network/zookeeper_proxy/decoder.h#L28). There are still several opcodes in [ZooDefs.java](https://github.com/apache/zookeeper/blob/89c1831f84891f425f1fa9224210587124f1c1ec/zookeeper-server/src/main/java/org/apache/zookeeper/ZooDefs.java#L36) but not in the previous list, I will try to support them in my next PR.

Reply:
Yeah, the names are consistent between the two.